src/playercards/AllCardsBag.ttslua

```diff
@@ -241,26 +520,119 @@ end
 --     Otherwise, a list of tables, each with the following fields
 --       cardData: TTS object data, suitable for spawning the card
 --       cardMetadata: Table of parsed metadata
-function getCardsByClassAndLevel(params)
+function getCardsByClass(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+  return getFilteredIndexList(classIndex[string.lower(params.class)], params.includeOthers)
+end
+
+function getCardsByCycle(params)
   if (not indexingDone) then
     broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
     return { }
   end
-  local upgradeKey
-  if (params.upgraded) then
-    upgradeKey = "-upgrade"
+  return getFilteredIndexList(cycleIndex[string.lower(params.cycleName)], params.includeOthers)
+end
+
+function getCardsByTrait(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+  return getFilteredIndexList(traitIndex[string.lower(params.trait)], params.includeOthers)
+end
+
+function getCardsByIcon(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+
+  return getFilteredIndexList(iconIndex[string.lower(params.icon)][tonumber(params.count)], params.includeOthers)
+end
+
+function getCardsByType(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+  return getFilteredIndexList(typeIndex[string.lower(params.cardType)], params.includeOthers)
+end
+
+function getCardsByCost(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+  return getFilteredIndexList(costIndex[tonumber(params.cost)], params.includeOthers)
+end
+
+function getCardsByLevel(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+  return getFilteredIndexList(levelIndex[tonumber(params.level)], params.includeOthers)
+end
+
+function getCardsByUses(params)
+  if (not indexingDone) then
+    broadcastToAll("Still loading player cards, please try again in a few seconds", {0.9, 0.2, 0.2})
+    return { }
+  end
+  if type(params.index) == "number" then
+    return getFilteredIndexList(usesIndex[tonumber(params.index)], params.includeOthers)
+  elseif usesIndex[string.lower(params.index)] then
+    return getFilteredIndexList(usesIndex[string.lower(params.index)], params.includeOthers)
```

Comment:
These functions should be turned into a single one that takes the respective index as parameter. If you end up repeating yourself this much, there is always a better approach.

Reply:
Agreed it does not look pretty as is.  What would be the best way to do what you're suggesting though?  Is the parameter a string and the single function is just one big chain of `elseif`s?